extracted_code/Z3Solver.ml

```diff
@@ -0,0 +1,41 @@
+open Z3
+
+(* Recursively convert a coq_SmtBoolExpr to a Z3 expression *)
+(* TODO: This function is trusted, so needs to be checked via other means like fuzzing *)
+let rec z3_expr_from_coq_smt_bool_expr (expr : SmtExpr.coq_SmtBoolExpr) (ctx : Z3.context)
+  : Z3.Expr.expr =
+  match expr with
+  | SmtExpr.SmtTrue -> Z3.Boolean.mk_true ctx
+  | SmtExpr.SmtFalse -> Z3.Boolean.mk_false ctx
+  | SmtExpr.SmtBoolAnd (e1, e2) -> Z3.Boolean.mk_and ctx [z3_expr_from_coq_smt_bool_expr e1 ctx; z3_expr_from_coq_smt_bool_expr e2 ctx]
+  | SmtExpr.SmtBoolOr (e1, e2) -> Z3.Boolean.mk_or ctx [z3_expr_from_coq_smt_bool_expr e1 ctx; z3_expr_from_coq_smt_bool_expr e2 ctx]
+  | SmtExpr.SmtBoolNot e -> Z3.Boolean.mk_not ctx (z3_expr_from_coq_smt_bool_expr e ctx)
+  | SmtExpr.SmtBoolEq (a1, a2) -> Z3.Boolean.mk_eq ctx (z3_expr_from_coq_smt_arith_expr a1 ctx) (z3_expr_from_coq_smt_arith_expr a2 ctx)
+and z3_expr_from_coq_smt_arith_expr (expr : SmtExpr.coq_SmtArithExpr) (ctx : Z3.context)
+  : Z3.Expr.expr =
+  match expr with
+  | SmtExpr.SmtConst n -> Z3.Arithmetic.Integer.mk_numeral_i ctx (Obj.magic n : int) (* TODO: Obj.magic is not type safe. *)
```

Comment:
I don't think we should be using (Obj.magic n) here. Instead, there should be a function available (in the extracted code) that converts n (from whatever type it is) into an int.